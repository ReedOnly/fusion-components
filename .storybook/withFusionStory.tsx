import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
