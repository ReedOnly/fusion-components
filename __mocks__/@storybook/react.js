import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
