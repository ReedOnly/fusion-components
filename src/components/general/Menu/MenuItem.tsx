import * as React from "react";
import classNames from "classnames";
import styles from "./styles.less";

export type MenuItemType = {
    key: string;
    title: string;
    aside?: string;
    isSelected?: boolean;
    isDisabled?: boolean;
};

type MenuItemProps = {
    item: MenuItemType;
    index: number;
    isFocused: boolean;
    onClick: (e: MenuItemType) => void;
};

const MenuItem: React.FC<MenuItemProps> = props => {
    const ref = React.useRef<HTMLButtonElement>(null);
    const onClick = e => {
        if (ref.current) {
            ref.current.blur();
        }

        props.onClick(props.item);
        e.preventDefault();
    };

    const className = classNames({
        [styles.isSelected]: props.item.isSelected,
        [styles.isFocused]: props.isFocused,
    });

    return (
        <button
            className={className}
            disabled={props.item.isDisabled}
            ref={ref}
            tabIndex={props.index + 1}
            onClick={onClick}
        >
            <span>
                <span>{props.item.title}</span>
                {props.item.aside && <aside>{props.item.aside}</aside>}
            </span>
        </button>
    );
};

export default MenuItem;
