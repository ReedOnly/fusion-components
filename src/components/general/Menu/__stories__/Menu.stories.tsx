import * as React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import withFusionStory from "../../../../../.storybook/withFusionStory";
import Menu from "../index";

const MenuStory = () => {
    const [ref, setRef] = React.useState<HTMLElement | null>(null);

    return (
        <React.Fragment>
        <input placeholder="Focus here to navigate" ref={setRef} />
            <Menu
                onClick={action("click")}
                keyboardNavigationRef={ref}
                sections={[
                    {
                        key: "This is the only section, but I still need a key",
                        items: [
                            {
                                key: "1",
                                title: "First",
                            },
                            {
                                key: "2",
                                title: "Selected",
                                isSelected: true,
                            },
                            {
                                key: "3",
                                title: "Disabled",
                                isDisabled: true,
                            },
                            {
                                key: "4",
                                title: "Last",
                            },
                        ],
                    },
                ]}
            />
        </React.Fragment>
    );
};

storiesOf("General|Menu", module)
    .addDecorator(withFusionStory("Menu"))
    .add("Default", () => <MenuStory />);
