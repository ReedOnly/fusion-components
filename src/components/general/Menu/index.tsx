import * as React from "react";
import classNames from "classnames";
import useElevationClassName, { Elevation } from "../../../hooks/useElevationClassName";
import useKeyboardNavigation from "../../../hooks/useKeyboardNavigation";
import MenuItem, { MenuItemType } from "./MenuItem";
import styles from "./styles.less";

export type MenuSection = {
    key: string,
    title?: string,
    items: MenuItemType[],
};

export type MenuProps = {
    sections: MenuSection[],
    elevation?: Elevation,
    onClick?: (item: MenuItemType) => void | Promise<void>,
    keyboardNavigationRef?: HTMLElement | null,
};

const Menu: React.FC<MenuProps> = ({ sections, elevation, onClick, keyboardNavigationRef }: MenuProps) => {
    const [focusedSectionKey, setFocusedSectionKey] = React.useState<string | null>(null);
    const [focusedItemKey, setFocusedItemKey] = React.useState<string | null>(null);

    const nextOrPrev = (direction: number) => {
        const sectionIndex = sections.findIndex(section => section.key === focusedSectionKey);
        let nextSectionIndex = sectionIndex;

        if (sectionIndex === -1) {
            nextSectionIndex = direction === -1 ? sections.length - 1 : 0;
        }

        const items = sections[nextSectionIndex].items;

        const itemIndex = items.findIndex(item => item.key === focusedItemKey);
        let nextItemIndex = itemIndex;

        if (nextItemIndex === -1) {
            nextSectionIndex = direction === -1 ? sections.length - 1 : 0;
            nextItemIndex = direction === -1 ? sections[nextSectionIndex].items.length - 1 : 0;
        } else if (nextItemIndex === 0 && direction === -1) {
            nextSectionIndex += (sectionIndex + sections.length - 1) % sections.length;
        } else if (nextItemIndex === items.length - 1 && direction === 1) {
            nextSectionIndex += (sectionIndex + sections.length + 1) % sections.length;
        } else {
            nextItemIndex += direction;
        }

        const nextSection = sections[nextSectionIndex];

        setFocusedSectionKey(nextSection.key);
        setFocusedItemKey(nextSection.items[nextItemIndex].key);

        if(nextSection.items[nextItemIndex].isDisabled) {
            nextOrPrev(direction);
        }
    };

    const reset = () => {
        setFocusedSectionKey(keyboardNavigationRef && sections.length ? sections[0].key : null);
        setFocusedItemKey(keyboardNavigationRef && sections.length && sections[0].items.length ? sections[0].items[0].key : null);
    };

    const onItemClick = item => {
        if (onClick) {
            onClick(item);
        }
    };

    useKeyboardNavigation({
        onUp: () => nextOrPrev(-1),
        onDown: () => nextOrPrev(+1),
        onEnter: () => {
            const section = sections.find(section => section.key === focusedSectionKey);
            if (!section) {
                return;
            }

            const item = section.items.find(item => item.key === focusedItemKey);
            if (!item) {
                return;
            }

            onItemClick(item);
        },
        onEscape: reset,
    }, keyboardNavigationRef);

    React.useEffect(reset, [sections]);

    const className = classNames(styles.container, useElevationClassName(elevation));

    return (
        <div className={className}>
            {sections.map(section => (
                <section key={section.key}>
                    {!!section.title && <h5>{section.title}</h5>}
                    {section.items.map((item, index) => (
                        <MenuItem
                            key={item.key}
                            index={index}
                            isFocused={
                                focusedSectionKey === section.key && focusedItemKey === item.key
                            }
                            item={item}
                            onClick={onItemClick}
                        />
                    ))}
                </section>
            ))}
        </div>
    );
};

Menu.defaultProps = {
    sections: [],
    elevation: 2,
};

export default Menu;
