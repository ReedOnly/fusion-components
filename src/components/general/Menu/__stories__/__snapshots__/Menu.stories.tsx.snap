// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|Menu Custom items 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container elevation2"
    >
      <section>
        <button
          className=""
          onClick={[Function]}
          tabIndex={1}
        >
          <span>
            <aside>
              <div
                className="container checkbox comfortable"
                onClick={[Function]}
              >
                <input
                  readOnly={true}
                  type="checkbox"
                />
                <label
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "borderColor": undefined,
                    }
                  }
                >
                  <span
                    className="checkmark"
                  />
                </label>
              </div>
            </aside>
            <span>
              <strong>
                1
              </strong>
               - 
              <em>
                Some option
              </em>
            </span>
          </span>
        </button>
        <button
          className=""
          onClick={[Function]}
          tabIndex={2}
        >
          <span>
            <aside>
              <div
                className="container checkbox comfortable"
                onClick={[Function]}
              >
                <input
                  readOnly={true}
                  type="checkbox"
                />
                <label
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "borderColor": undefined,
                    }
                  }
                >
                  <span
                    className="checkmark"
                  />
                </label>
              </div>
            </aside>
            <span>
              <strong>
                2
              </strong>
               - 
              <em>
                This is another option
              </em>
            </span>
          </span>
        </button>
        <button
          className=""
          onClick={[Function]}
          tabIndex={3}
        >
          <span>
            <aside>
              <div
                className="container checkbox comfortable"
                onClick={[Function]}
              >
                <input
                  readOnly={true}
                  type="checkbox"
                />
                <label
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "borderColor": undefined,
                    }
                  }
                >
                  <span
                    className="checkmark"
                  />
                </label>
              </div>
            </aside>
            <span>
              <strong>
                3
              </strong>
               - 
              <em>
                You can check this one 
                <svg
                  height={24}
                  style={
                    Object {
                      "color": "",
                    }
                  }
                  viewBox="0 0 24 24"
                  width={24}
                >
                  <path
                    clipRule="evenodd"
                    d="M8.80004 15.905L5.30004 12.405C5.11519 12.2179 4.86309 12.1125 4.60004 12.1125C4.33699 12.1125 4.0849 12.2179 3.90004 12.405C3.51004 12.795 3.51004 13.415 3.90004 13.805L8.09004 17.995C8.48004 18.385 9.11004 18.385 9.50004 17.995L20.1 7.40501C20.49 7.01501 20.49 6.39501 20.1 6.00501C19.9152 5.81786 19.6631 5.71252 19.4 5.71252C19.137 5.71252 18.8849 5.81786 18.7 6.00501L8.80004 15.905Z"
                    fill="currentColor"
                    fillRule="evenodd"
                  />
                </svg>
              </em>
            </span>
          </span>
        </button>
        <button
          className=""
          disabled={true}
          onClick={[Function]}
          tabIndex={4}
        >
          <span>
            <aside>
              <div
                className="container checkbox comfortable disabled"
                onClick={[Function]}
              >
                <input
                  disabled={true}
                  readOnly={true}
                  type="checkbox"
                />
                <label
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "borderColor": undefined,
                    }
                  }
                >
                  <span
                    className="checkmark"
                  />
                </label>
              </div>
            </aside>
            <span>
              <strong>
                4
              </strong>
               - 
              <em>
                This one is disabled
              </em>
            </span>
          </span>
        </button>
      </section>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`General|Menu Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <input
      placeholder="Focus here to navigate"
    />
    <div
      className="container elevation2"
    >
      <section>
        <button
          className=""
          onClick={[Function]}
          tabIndex={1}
        >
          <span>
            <span>
              First
            </span>
          </span>
        </button>
        <button
          className="isSelected"
          onClick={[Function]}
          tabIndex={2}
        >
          <span>
            <span>
              Selected
            </span>
          </span>
        </button>
        <button
          className=""
          disabled={true}
          onClick={[Function]}
          tabIndex={3}
        >
          <span>
            <span>
              Disabled
            </span>
          </span>
        </button>
        <button
          className=""
          onClick={[Function]}
          tabIndex={4}
        >
          <span>
            <span>
              Last
            </span>
          </span>
        </button>
      </section>
    </div>
    <div
      style={
        Object {
          "width": "300px",
        }
      }
    >
      <input
        placeholder="Narrow menu"
      />
      <div
        className="container elevation2"
      >
        <section>
          <button
            className=""
            onClick={[Function]}
            tabIndex={1}
          >
            <span>
              <span>
                First
              </span>
            </span>
          </button>
          <button
            className="isSelected"
            onClick={[Function]}
            tabIndex={2}
          >
            <span>
              <span>
                Selected
              </span>
            </span>
          </button>
          <button
            className=""
            disabled={true}
            onClick={[Function]}
            tabIndex={3}
          >
            <span>
              <span>
                Disabled
              </span>
            </span>
          </button>
          <button
            className=""
            onClick={[Function]}
            tabIndex={4}
          >
            <span>
              <span>
                Last
              </span>
            </span>
          </button>
        </section>
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
